finetuner/tuner/callback/training_checkpoint.py

```diff
@@ -0,0 +1,166 @@
+import os
+import pickle
+from typing import TYPE_CHECKING
+import shutil
+
+import keras
+import paddle
+import torch
+from jina.logging.logger import JinaLogger
+
+from finetuner.helper import get_framework
+
+from .base import BaseCallback
+
+if TYPE_CHECKING:
+    from ..base import BaseTuner
+
+
+class TrainingCheckpoint(BaseCallback):
+    """
+    Callback to save model at every epoch or the last K epochs
+    """
+
+    def __init__(self, save_dir: str, last_k_epochs: int = None):
+        """
+        :param save_dir: string, path to save the model file.
+        :param last_k_epochs: this parameter is an integer. It allows you to save
+            on only the most recent epochs and not on every epoch.
+            For example if `last_k_epochs = 3` and the total number of epochs is 10
+            epoch 5 -> saved 5, 4, 3
+            epoch 6 -> saved 6, 5, 4 (here saved epoch 3 is deleted)
+            epoch 7 -> saved 7, 6, 5 (here saved epoch 4 is deleted)
+        """
+        self._logger = JinaLogger(self.__class__.__name__)
+        self._save_dir = save_dir
+        self._last_k_epochs = last_k_epochs
+        if not save_dir:
+            raise ValueError(
+                '``save_dir`` parameter is mandatory. Pass it in parameters'
+            )
+
+    def on_epoch_end(self, tuner: 'BaseTuner'):
+        """
+        Called at the end of the training epoch.
+        """
+        self._save_model_framework(tuner)
+        self._logger.logger.info(
+            f'Model trained for {tuner.state.epoch+1} epochs is saved!'
+        )
+        if self._last_k_epochs:
```

Comment:
I think this part is not correctly implemented, and in any case way more lines than needed. Here's how I think this (deleting of excess checkpoints) could be done:

- create a property `saved_checkpoints`, which is a list with paths to saved checkpoints
- when creating a new checkpoints, append to the list
- when checking for deletion: check if `len(saved_checkpoints) > last_k_epochs` - if so, take the first element from `saved_checkpoints`, and delete the file, and after that remove the path from ``saved_checkpoints`